src/boot/efi/stub.c

```diff
@@ -180,6 +181,153 @@ static bool use_load_options(
         return true;
 }
 
+static EFI_STATUS get_cmdline_mule(
+                EFI_HANDLE stub_image,
+                EFI_LOADED_IMAGE_PROTOCOL *loaded_image,
+                char16_t **ret_buffer) {
+
+        _cleanup_(file_closep) EFI_FILE *root = NULL, *extra_dir = NULL;
+        size_t dirent_size = 0, n_items = 0, n_allocated = 0;
+        _cleanup_free_ EFI_FILE_INFO *dirent = NULL, *extra_dir_info = NULL;
+        _cleanup_(strv_freep) char16_t **items = NULL;
+        _cleanup_free_ char16_t *buffer = NULL;
+        EFI_STATUS err;
+
+        assert(stub_image);
+        assert(loaded_image);
+        assert(ret_buffer);
+
+        if (!loaded_image->DeviceHandle)
+                goto nothing;
+
+        err = open_volume(loaded_image->DeviceHandle, &root);
+        if (err == EFI_UNSUPPORTED)
+                /* Error will be unsupported if the bootloader doesn't implement the file system protocol on
+                 * its file handles. */
+                goto nothing;
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Unable to open root directory: %m");
+
+        err = open_directory(root, u"\\loader\\mules", &extra_dir); // TODO: bikeshedding
+        if (err == EFI_NOT_FOUND)
+                /* No extra subdir, that's totally OK */
+                goto nothing;
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to open mules directory: %m");
+
+        err = get_file_info(extra_dir, &extra_dir_info, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to query information of mules directory: %m");
+
+        for (;;) {
+                _cleanup_free_ char16_t *d = NULL;
+
+                err = readdir(extra_dir, &dirent, &dirent_size);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to read extra directory of loaded image: %m");
+                if (!dirent) /* End of directory */
+                        break;
+
+                if (dirent->FileName[0] == '.')
+                        continue;
+                if (FLAGS_SET(dirent->Attribute, EFI_FILE_DIRECTORY))
+                        continue;
+                if (!endswith_no_case(dirent->FileName, u".mule.cmdline.efi")) // TODO: bikeshedding
```

Comment:
hmm, these arne't real efi programs, i'd use a different suffix hence. they are efi programs to the same level as windows exe files are efi programs: blobs in PE format i guess.

Also, I think i wouldn't limit this by name to cmdlines already. I am pretty sure people want to use this for other stuff that can be embedded in a UKI too, i.e. devicetree stuff, extra initrds and so on.

Reply:
Yes, there's a bunch of such considerations to make, including for example giving different semantics meaning - ie, a signed and verified credential addon should be treated differently from an unsigned in some way, for example loading by default vs not and so on.

But before getting bogged down in all that, I want to get the mule mechanism fully sorted first, given it requires shim changes. I'll keep this as draft until that is sorted.

Reply:
I'll switch the suffix to .cmdline - I think it makes sense to have, to filter out early and avoid parsing and validating needlessly through shim.

I'll also move to EFI\systemd\addons as we already have EFI\systemd\drivers. We need to figure out if we want a subdirectory per type, or filename suffix, to differentiate.

Reply:
mmh so currently we load:

EFI/systemd/drivers/*
EFI/Linux/foo.efi.extra.d/*.cred
loader/credentials/*.cred

This is a bit confusing... so we definitely want the .cmdline suffix to eventually store them in foo.efi.extra.d/ and they would only apply to that UKI. Where do we store the global cmdline? EFI/systemd/cmdlines/ ? or loader/cmdlines/?